### ISSUE TITLE
Is there a connection limit?

### ISSUE THREAD

**Issue by caspervonb** [Microsoft/edge-diagnostics-adapter#16]
Does the adaptor support connections per debugging target? Might be worth documenting somewhere.

**Comment by andysterland** [comment]
The web socket end point for each page can have multiple connections. Though the clients would likely hit issues and it would be unlikely to work off the bat. Having multiple remote clients wasn't something we had in mind to support. Though you can still use the F12 dev tools in Edge (except debugging) as they connect via a different mechanism. 

Is there a scenario you had in mind? I can sure imagine their being scenarios for having multiple tools working against the same page, the trick is not having them trip over each other.

**Comment by caspervonb** [comment]
> The web socket end point for each page can have multiple connections. Though the clients would likely hit issues and it would be unlikely to work off the bat. Having multiple remote clients wasn't something we had in mind to support. Though you can still use the F12 dev tools in Edge (except debugging) as they connect via a different mechanism.

There are basically only two criterias I can think of to consider it *supported*, Chrome still has an [open issue](https://bugs.chromium.org/p/chromium/issues/detail?id=129539) regarding this as it only supports a single connection. Other adapters seem to favor supporting multiple connections however.

* Enable/disable state per connection (e.g enable should send all relevant buffered notifications to that client)
* Per connection command mapping (e.g command identifiers are local to a client)

With these two taken care of, most clients will continue on as normal as they are just sending commands as normal and receiving events as normal.

> Is there a scenario you had in mind? I can sure imagine their being scenarios for having multiple tools working against the same page, the trick is not having them trip over each other.

Basically *anything* that is non-trivial, for example running [coda-hotpatch](https://github.com/codajs/coda-hotpatch) concurrently with a graphical debugger open like [vscode-chrome-debug](https://github.com/Microsoft/vscode-chrome-debug), or just vanilla chrome developer tools.

**Comment by andysterland** [comment]
Thanks for the follow up :). You're dead right there is a lot of interesting scenarios for multiple clients indeed it's something we've been playing around with. 

What I should have said, but didn't, is we don't have plans to support it right now and by support I mean test it and fix issues :). Our first stop on the roadmap is to get the adapter to the state in which can support debugging clients like VS Code then fill out some of the APIs that are missing/buggy. And my initial reply was in that somewhat single minded short term hat on :(. In the future for sure we absolutely want to have many clients connected to the same page.

If you're interested it be great to get coda-hotpatch working against the adapter it looks like it uses:
1. 'CSS.styleSheetAdded'
2. 'Debugger.setScriptSource'
3. 'CSS.styleSheetAdded'
4. 'Debugger.globalObjectCleared'
5. 'Debugger.scriptParsed'
The adapter doesn't yet implement them all but they are all on the short term to do list as the platform in Edge has the capabilities (though setScriptSource only works when not at a breakpoint).

**Comment by caspervonb** [comment]
> What I should have said, but didn't, is we don't have plans to support it right now and by support I mean test it and fix issues ...

Noted, but what are the *current* semantics?

* Are command identifiers local per client or do they need to be unique across all clients?
* Is enabled/disabled state (event delivery) tracked per client, or per target?

> If you're interested it be great to get coda-hotpatch working against the adapter it looks like it uses ...

Working towards that :wink:

**Comment by molant** [comment]
@caspervonb the [Chrome issue you reference is now fixed](https://bugs.chromium.org/p/chromium/issues/detail?id=129539).

Closing this issue as it is quite old and there are no plan to add that feature to this project.
